File: backend/app/api/api_v1/endpoints/projects.py
"""
项目管理API端点
"""
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func, or_

from app.db.database import get_db
from app.core.deps import get_current_active_user
from app.models.user import User
from app.models.project import Project
from app.services.project import ProjectService
from app.schemas.project import (
    CreateProjectRequest, UpdateProjectRequest, ProjectResponse,
    ProjectListParams, ProjectListResponse, ProjectStats,
    NameAvailabilityResponse, SearchProjectResponse, TaskResponse,
    ProjectEnvironmentVariablesUpdate, ProjectEnvDetectionResponse
)

router = APIRouter()


@router.get("/", response_model=ProjectListResponse)
async def get_projects(
    page: int = Query(1, ge=1, description="页码"),
    limit: int = Query(10, ge=1, le=100, description="每页数量"),
    search: Optional[str] = Query(None, max_length=100, description="搜索关键词"),
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取项目列表
    
    支持分页、搜索、筛选和排序功能。
    """
    # 构建基础查询 - 只返回用户的项目
    query = select(Project).where(Project.user_id == current_user.id)
    
    # 添加搜索条件
    if search:
        search_condition = or_(
            Project.name.ilike(f"%{search}%"),
            Project.description.ilike(f"%{search}%"),
            Project.git_url.ilike(f"%{search}%")
        )
        query = query.where(search_condition)
    
    # 获取总数
    count_query = select(func.count()).select_from(query.subquery())
    total_result = await db.execute(count_query)
    total = total_result.scalar()
    
    # 添加分页和排序
    offset = (page - 1) * limit
    query = query.order_by(Project.updated_at.desc()).offset(offset).limit(limit)
    
    # 执行查询
    result = await db.execute(query)
    projects = result.scalars().all()
    
    # 计算页数
    pages = (total + limit - 1) // limit
    
    return ProjectListResponse(
        items=[ProjectResponse.model_validate(project) for project in projects],
        total=total,
        page=page,
        limit=limit,
        pages=pages
    )


@router.post("/", response_model=ProjectResponse, status_code=status.HTTP_201_CREATED)
async def create_project(
    project_data: CreateProjectRequest,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    创建新项目
    
    创建一个新的项目，用户必须已登录。
    """
    project = await ProjectService.create_project(db, project_data, current_user.id)
    return ProjectResponse.from_orm(project)


@router.get("/stats", response_model=ProjectStats)
async def get_project_stats(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取项目统计数据
    
    返回用户的项目和任务统计信息。
    """
    return await ProjectService.get_project_stats(db, current_user.id)


@router.get("/recent", response_model=List[ProjectResponse])
async def get_recent_projects(
    limit: int = Query(10, ge=1, le=50, description="返回数量限制"),
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取最近使用的项目
    
    返回用户最近活跃的项目列表。
    """
    projects = await ProjectService.get_recent_projects(db, current_user.id, limit)
    return [ProjectResponse.from_orm(project) for project in projects]


@router.get("/search", response_model=SearchProjectResponse)
async def search_projects(
    q: str = Query(..., min_length=1, max_length=100, description="搜索关键词"),
    limit: int = Query(10, ge=1, le=50, description="返回数量限制"),
    filter_status: Optional[str] = Query(None, description="状态筛选"),
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    搜索项目
    
    根据项目名称和描述进行模糊搜索。
    """
    filters = {}
    if filter_status:
        filters["status"] = filter_status
    
    projects, total = await ProjectService.search_projects(
        db, current_user.id, q, limit, filters
    )
    
    return SearchProjectResponse(
        items=[ProjectResponse.from_orm(project) for project in projects],
        total=total,
        query=q
    )


@router.get("/check-name", response_model=NameAvailabilityResponse)
async def check_project_name(
    name: str = Query(..., min_length=1, max_length=100, description="项目名称"),
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    检查项目名称是否可用
    
    返回名称可用性和建议的替代名称。
    """
    available, suggestions = await ProjectService.check_name_availability(
        db, name, current_user.id
    )
    
    return NameAvailabilityResponse(
        available=available,
        suggestions=suggestions
    )


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取单个项目详情
    
    返回指定项目的详细信息。用户只能访问自己的项目。
    """
    project = await ProjectService.get_project(db, project_id, current_user.id)
    return ProjectResponse.from_orm(project)


@router.put("/{project_id}", response_model=ProjectResponse)
async def update_project(
    project_id: int,
    project_data: UpdateProjectRequest,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    更新项目
    
    更新指定项目的信息。用户只能更新自己的项目。
    """
    project = await ProjectService.update_project(
        db, project_id, current_user.id, project_data
    )
    return ProjectResponse.from_orm(project)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    删除项目
    
    永久删除指定项目及其相关数据。此操作不可逆。
    """
    await ProjectService.delete_project(db, project_id, current_user.id)


@router.get("/{project_id}/stats", response_model=ProjectStats)
async def get_single_project_stats(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取单个项目的统计数据
    
    返回指定项目的任务统计信息。
    """
    return await ProjectService.get_project_stats(db, current_user.id, project_id)


@router.post("/{project_id}/clone", response_model=TaskResponse)
async def clone_repository(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    克隆Git仓库
    
    为指定项目克隆Git仓库。返回异步任务ID。
    """
    # TODO: 实现Git克隆功能
    return TaskResponse(
        task_id=f"clone-{project_id}-{current_user.id}",
        status="pending"
    )


@router.post("/{project_id}/pull", response_model=TaskResponse)
async def pull_repository(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    拉取最新代码
    
    为指定项目拉取Git仓库的最新代码。返回异步任务ID。
    """
    # TODO: 实现Git拉取功能
    return TaskResponse(
        task_id=f"pull-{project_id}-{current_user.id}",
        status="pending"
    )


@router.post("/{project_id}/commands")
async def execute_command(
    project_id: int,
    command: Dict[str, Any],
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    执行Claude CLI命令
    
    在指定项目中执行Claude CLI命令。
    """
    # TODO: 实现命令执行功能
    return {
        "message": f"Execute command for project {project_id}",
        "project_id": project_id,
        "user_id": current_user.id,
        "command": command
    }


@router.get("/check-name", response_model=NameAvailabilityResponse)
async def check_project_name_availability(
    name: str = Query(..., min_length=1, max_length=100, description="项目名称"),
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """
    检查项目名称是否可用
    
    检查指定名称是否已被当前用户使用，如果不可用，提供替代建议。
    """
    return await ProjectService.check_name_availability(db, current_user.id, name)


@router.get("/{project_id}/environment", response_model=dict)
async def get_project_environment_variables(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """获取项目环境变量"""
    project = await ProjectService.get_project(db, project_id, current_user.id)
    return {
        "environment_variables": project.environment_variables or {},
        "detected_env_vars": project.detected_env_vars or {}
    }


@router.put("/{project_id}/environment", response_model=dict)
async def update_project_environment_variables(
    project_id: int,
    env_update: ProjectEnvironmentVariablesUpdate,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """更新项目环境变量"""
    project = await ProjectService.get_project(db, project_id, current_user.id)
    project.environment_variables = env_update.environment_variables
    
    await db.commit()
    await db.refresh(project)
    
    return {
        "environment_variables": project.environment_variables or {},
        "message": "环境变量更新成功"
    }


@router.post("/{project_id}/detect-env", response_model=ProjectEnvDetectionResponse)
async def detect_project_environment_variables(
    project_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """AI自动检测项目环境变量"""
    # TODO: 实现AI环境变量检测功能
    # 这里先返回模拟数据
    from app.schemas.project import DetectedEnvVar
    
    mock_detected_vars = [
        DetectedEnvVar(
            name="DATABASE_URL",
            description="数据库连接URL",
            default_value="postgresql://localhost:5432/mydb",
            source_file="config/database.py",
            line_number=15,
            required=True,
            category="database"
        ),
        DetectedEnvVar(
            name="SECRET_KEY",
            description="应用密钥",
            source_file="app/settings.py",
            line_number=8,
            required=True,
            category="security"
        ),
        DetectedEnvVar(
            name="DEBUG",
            description="调试模式开关",
            default_value="False",
            source_file="app/settings.py",
            line_number=12,
            required=False,
            category="general"
        )
    ]
    
    return ProjectEnvDetectionResponse(
        detected_vars=mock_detected_vars,
        total_found=len(mock_detected_vars),
        source_files=["config/database.py", "app/settings.py"],
        suggestions=[
            "建议为所有数据库相关的环境变量设置默认值",
            "SECRET_KEY应该使用强随机字符串",
            "生产环境中请确保DEBUG设置为False"
        ]
    )

File: frontend/src/api/projects.ts
/**
 * 项目相关API
 */
import { apiClient } from '@/lib/api'

// 类型定义
export interface Project {
  id: string
  name: string
  description?: string
  git_url?: string
  branch?: string
  status: 'active' | 'inactive' | 'archived'
  owner_id: string
  created_at: string
  updated_at: string
  last_activity?: string
  settings?: ProjectSettings
}

export interface ProjectSettings {
  auto_deploy: boolean
  notification_enabled: boolean
  environment_variables: Record<string, string>
  build_command?: string
  test_command?: string
}

export interface CreateProjectRequest {
  name: string
  description?: string
  git_url?: string
  branch?: string
  source?: 'github' | 'gitlab' | 'manual'
  github_repo_id?: number
  settings?: Partial<ProjectSettings>
}

export interface UpdateProjectRequest {
  name?: string
  description?: string
  git_url?: string
  branch?: string
  status?: Project['status']
  settings?: Partial<ProjectSettings>
}

export interface ProjectListParams {
  page?: number
  limit?: number
  search?: string
  status?: Project['status']
  owner_id?: string
  sort_by?: 'name' | 'created_at' | 'updated_at' | 'last_activity'
  sort_order?: 'asc' | 'desc'
}

export interface ProjectListResponse {
  items: Project[]
  total: number
  page: number
  limit: number
  pages: number
}

export interface ProjectStats {
  total_projects: number
  active_projects: number
  total_tasks: number
  successful_tasks: number
  failed_tasks: number
  success_rate: number
}

export interface ProjectEnvironmentVariables {
  environment_variables: Record<string, string>
  detected_env_vars: Record<string, any>
}

export interface ProjectEnvironmentUpdate {
  environment_variables: Record<string, string>
}

export interface DetectedEnvVar {
  name: string
  description?: string
  default_value?: string
  source_file: string
  line_number: number
  required: boolean
  category: string
}

export interface ProjectEnvDetectionResponse {
  detected_vars: DetectedEnvVar[]
  total_found: number
  source_files: string[]
  suggestions: string[]
}

// 项目API服务类
export class ProjectApiService {
  private static readonly BASE_PATH = '/api/v1/projects'

  /**
   * 获取项目列表
   */
  static async getProjects(params?: ProjectListParams): Promise<ProjectListResponse> {
    const queryParams = new URLSearchParams()
    
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, value.toString())
        }
      })
    }

    const url = queryParams.toString() 
      ? `${this.BASE_PATH}?${queryParams.toString()}`
      : this.BASE_PATH

    return apiClient.get<ProjectListResponse>(url)
  }

  /**
   * 获取单个项目详情
   */
  static async getProject(id: string): Promise<Project> {
    return apiClient.get<Project>(`${this.BASE_PATH}/${id}`)
  }

  /**
   * 创建新项目
   */
  static async createProject(data: CreateProjectRequest): Promise<Project> {
    return apiClient.post<Project>(this.BASE_PATH, data)
  }

  /**
   * 更新项目
   */
  static async updateProject(id: string, data: UpdateProjectRequest): Promise<Project> {
    return apiClient.put<Project>(`${this.BASE_PATH}/${id}`, data)
  }

  /**
   * 删除项目
   */
  static async deleteProject(id: string): Promise<void> {
    await apiClient.delete(`${this.BASE_PATH}/${id}`)
  }

  /**
   * 归档项目
   */
  static async archiveProject(id: string): Promise<Project> {
    return apiClient.patch<Project>(`${this.BASE_PATH}/${id}/archive`)
  }

  /**
   * 恢复项目
   */
  static async restoreProject(id: string): Promise<Project> {
    return apiClient.patch<Project>(`${this.BASE_PATH}/${id}/restore`)
  }

  /**
   * 克隆Git仓库
   */
  static async cloneRepository(id: string): Promise<{
    task_id: string
    status: string
  }> {
    return apiClient.post<{
      task_id: string
      status: string
    }>(`${this.BASE_PATH}/${id}/clone`)
  }

  /**
   * 拉取最新代码
   */
  static async pullRepository(id: string): Promise<{
    task_id: string
    status: string
  }> {
    return apiClient.post<{
      task_id: string
      status: string
    }>(`${this.BASE_PATH}/${id}/pull`)
  }

  /**
   * 获取项目统计数据
   */
  static async getProjectStats(id?: string): Promise<ProjectStats> {
    const url = id ? `${this.BASE_PATH}/${id}/stats` : `${this.BASE_PATH}/stats`
    return apiClient.get<ProjectStats>(url)
  }

  /**
   * 获取项目活动历史
   */
  static async getProjectActivity(id: string, params?: {
    page?: number
    limit?: number
    type?: string
  }): Promise<{
    items: Array<{
      id: string
      type: string
      description: string
      created_at: string
      metadata?: Record<string, any>
    }>
    total: number
    page: number
    limit: number
  }> {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, value.toString())
        }
      })
    }

    const url = queryParams.toString()
      ? `${this.BASE_PATH}/${id}/activity?${queryParams.toString()}`
      : `${this.BASE_PATH}/${id}/activity`

    return apiClient.get(url)
  }

  /**
   * 获取项目文件树
   */
  static async getProjectFiles(id: string, path: string = ''): Promise<{
    files: Array<{
      name: string
      path: string
      type: 'file' | 'directory'
      size?: number
      modified_at: string
    }>
    current_path: string
  }> {
    const queryParams = path ? `?path=${encodeURIComponent(path)}` : ''
    return apiClient.get(`${this.BASE_PATH}/${id}/files${queryParams}`)
  }

  /**
   * 获取文件内容
   */
  static async getFileContent(id: string, filePath: string): Promise<{
    content: string
    encoding: string
    size: number
    modified_at: string
  }> {
    return apiClient.get(`${this.BASE_PATH}/${id}/files/content`, {
      params: { path: filePath }
    })
  }

  /**
   * 更新项目设置
   */
  static async updateProjectSettings(
    id: string, 
    settings: Partial<ProjectSettings>
  ): Promise<ProjectSettings> {
    return apiClient.put<ProjectSettings>(`${this.BASE_PATH}/${id}/settings`, settings)
  }

  /**
   * 获取项目环境变量
   */
  static async getEnvironmentVariables(id: string): Promise<ProjectEnvironmentVariables> {
    return apiClient.get(`${this.BASE_PATH}/${id}/environment`)
  }

  /**
   * 更新项目环境变量
   */
  static async updateEnvironmentVariables(
    id: string, 
    variables: ProjectEnvironmentUpdate
  ): Promise<{ environment_variables: Record<string, string>; message: string }> {
    return apiClient.put(`${this.BASE_PATH}/${id}/environment`, variables)
  }

  /**
   * AI自动检测项目环境变量
   */
  static async detectEnvironmentVariables(id: string): Promise<ProjectEnvDetectionResponse> {
    return apiClient.post(`${this.BASE_PATH}/${id}/detect-env`, {})
  }

  /**
   * 搜索项目
   */
  static async searchProjects(query: string, params?: {
    limit?: number
    filters?: {
      status?: Project['status']
      owner_id?: string
    }
  }): Promise<{
    items: Project[]
    total: number
    query: string
  }> {
    const searchParams = new URLSearchParams({ q: query })
    
    if (params?.limit) {
      searchParams.append('limit', params.limit.toString())
    }
    
    if (params?.filters) {
      Object.entries(params.filters).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          searchParams.append(`filter_${key}`, value.toString())
        }
      })
    }

    return apiClient.get(`${this.BASE_PATH}/search?${searchParams.toString()}`)
  }

  /**
   * 获取我的项目
   */
  static async getMyProjects(params?: Omit<ProjectListParams, 'owner_id'>): Promise<ProjectListResponse> {
    return this.getProjects({ ...params, owner_id: 'me' })
  }

  /**
   * 获取最近使用的项目
   */
  static async getRecentProjects(limit: number = 10): Promise<Project[]> {
    return apiClient.get(`${this.BASE_PATH}/recent?limit=${limit}`)
  }

  /**
   * 检查项目名称是否可用
   */
  static async checkProjectNameAvailability(name: string): Promise<{
    available: boolean
    suggestions?: string[]
  }> {
    return apiClient.get(`${this.BASE_PATH}/check-name`, {
      params: { name }
    })
  }
}

export default ProjectApiService